### ISSUE TITLE
[api-extractor] ES modules and default import support

### ISSUE THREAD

**Issue by jonathandturner** [microsoft/web-build-tools#1264]
Currently in TypeScript, if you're using ES6 modules and you use default exports, you can access the default import via a call like this:

```ts
import localName from 'my_lib';
```

As of 7.1.1, api-extractor would turn this into the invalid code:

```ts
import { default } from 'my_lib';
```

To work around this, I tried a few things:

Rewriting the above to "import localName = require('my_lib')" doesn't work because TypeScript doesn't support this form for ES6 modules.

Using "import * as localName from 'my_lib' " doesn't have the same invalid code issue, but unfortunately it changes the API surface and isn't the recommended way of working with ES6 default imports.

**Comment by jonathandturner** [comment]
Working on a PR for this, which I hope to post today.

**Comment by jonathandturner** [comment]
PR is up now at https://github.com/microsoft/web-build-tools/pull/1267

**Comment by octogonz** [comment]
Thanks for implementing this! I'll try to look at it soon.

**Comment by octogonz** [comment]
By the way, I've had trouble with the `esModuleInterop` setting recently as people start to use it.  When a library has this setting enabled, it seems to break other packages that try to consume it, unless they also enable `esModuleInterop` .  Otherwise there is no way to import the symbols from that library.

Does that fit with your understanding of this feature?  If so, a library would want to procrastinate enabling this feature as long as possible, to ensure maximum compatibility.

@iclanton

**Comment by acchou** [comment]
I had recent experience with `esModuleInterop`. As I understand it, it shouldn't affect how a library is consumed, only how it imports other modules. The only breakage from exporting from a library appears to be this issue, when api-extractor generates a broken .d.ts rollup file.

This blog from 2015 (!) helps clarify what Babel does: http://2ality.com/2015/12/babel-commonjs.html. `esModuleInterop` is doing something very similar for TypeScript. One scenario is when enabling `esModuleInterop`, doing `import * as X from Y` stops working if X is a function or class (and of course if Y is a commonjs module, because es6 modules don't allow anything other than an object for X). The fix is to use `import X from Y` instead, when importing non-object values from commonjs modules.

**Comment by octogonz** [comment]
@acchou interestingly in https://github.com/microsoft/web-build-tools/pull/1267#discussion_r281850770 @jonathandturner seems to be reporting exactly what I said: once `esModuleInterop` is enabled, all dependent projects must enable it also, or else they fail to build. That comment provides a straightforward repro.

**Comment by jonathandturner** [comment]
@octogonz - this seems to happen because the .d.ts file created by the first module needs to be read by the second. In effect, you're outputting something that has `esModuleInterop` mode enabled, so it'd make sense that you'd need to enable support to consume the .d.ts file.

You could possibly avoid this by not using the technique when you're re-exporting the symbol, if you wanted to maximize compatibility.

**Comment by octogonz** [comment]
Does the problem only arise when you do `export { X } from "y";`?  

Or would it also be a problem if you did `import { X } from "y";` and then exported a class that inherits from `X`?

**Comment by acchou** [comment]
Actually, I don't think enabling `esModuleInterop` should force consumers to enable the flag. Can you post the compiler error message?

**Comment by jonathandturner** [comment]
@octogonz - it's around the default imports, and then exporting something using those default imports later.  Something like this:

```ts
import Long from "long";
export declare class UseLong {
  use_long(): Long;
}
```

@acchou - If the default import uses the style that requires `esModuleInterop` to be able to parse, it's not going to work.

Here's the error:

```
---> node node_modules/typescript/lib/tsc
../api-extractor-test-01/dist/api-extractor-test-01.d.ts(14,8): error TS1192: Module '"C:/Source/web-build-tools/common/temp/node_modules/.registry.npmjs.org/@types/long/4.0.0/node_modules/@types/long/index"' has no default export.
```

This error goes away when you enable `esModuleInterop`.

**Comment by jonathandturner** [comment]
Specifically the `import Long from "long";` fails to work.

**Comment by octogonz** [comment]
Anyway this seems to support the original conclusion:

> If so, a library project should want to procrastinate enabling this feature as long as possible, to ensure maximum compatibility with all possible consumers.

...which goes against the recommendations from the compiler docs.  But perhaps they didn't consider the compatibility goals of a shared library.  Or perhaps they assumed that enabling the feature is trivially easy.

**Comment by octogonz** [comment]
PR #1267 was released with API Extractor 7.1.5
